### PR TITLE
T19079

### PR DESCRIPTION
**Pull request by wgreenberg**
https://phabricator.endlessm.com/T19079

### CODE REVIEW

File: test/lib/index.test.js [hunk 1]
```diff
@@ -2,15 +2,50 @@
...
+        if (this._err) {
+            return Promise.reject(this._err);
+        } else {
+            return Promise.resolve();
+        }
+    }
+}
+
+function rmrf (dir) {
+    fs.readdirSync(dir).forEach(file => fs.unlinkSync(`${dir}/${file}`));
```
Comment (sgnn7) [inline test/lib/index.test.js]: `fs-extra` could be easier but not a blocker (https://github.com/jprichardson/node-fs-extra/blob/master/lib/remove/__tests__/remove-dir.test.js)

File: test/lib/index.test.js [hunk 2]
```diff
@@ -2,15 +2,50 @@
...
+        if (this._dependents) {
+            return this._dependents.map(asset => asset.asset_id);
+        } else {
+            return [];
+        }
+    }
+    _process() {
+        if (this._err) {
+            return Promise.reject(this._err);
+        } else {
```
Comment (sgnn7) [inline test/lib/index.test.js]: No need for else

File: test/lib/index.test.js [hunk 3]
```diff
@@ -2,15 +2,50 @@
...
+    },
+});
+
+class MockAsset extends libingester.BaseAsset {
+    fails_with_error(err) { this._err = err; }
+    set_dependent_assets(dependents) { this._dependents = dependents; }
+    get_dependent_assets() {
+        if (this._dependents) {
+            return this._dependents.map(asset => asset.asset_id);
+        } else {
```
Comment (sgnn7) [inline test/lib/index.test.js]: No need for else

File: lib/index.js [hunk 4]
```diff
@@ -824,6 +830,47 @@ class Hatch {
...
+
+        // In a second pass, remove all otherwise successful assets which had
+        // failed siblings
+        successful_assets = successful_assets.filter(asset => {
+            return !dependents_to_prune.includes(asset.asset_id);
+        });
+
+        // If this hatch had 100% failures, consider it failed and refuse to
+        // continue
+        if (successful_assets.length === 0 && this._assets.length > 0) {
```
Comment (sgnn7) [inline lib/index.js]: Maybe this can be set from the caller rather than assuming that any % of failures is fine?

Comment (wgreenberg) [reply lib/index.js]: maybe if that turns out to be necessary, but tbh I don't really see a use case for it

File: test/lib/index.test.js (hunk 1 above)
Comment (wgreenberg) [reply test/lib/index.test.js]: IMO the fewer dependencies the better

**Comment by wgreenberg** [comment]
@sgnn7 fixed up

File: lib/index.js (hunk 4 above)
Comment (sgnn7) [reply lib/index.js]: I think there's somewhat of an issue if 99% of articles fail - maybe we can even hard-code some sensible level here?

**Comment by wgreenberg** [comment]
@sgnn7 added the max fail rate

**Comment by sgnn7** [comment]
@wgreenberg LGTM - just needs an addition of version bump

**Comment by wgreenberg** [comment]
@sgnn7 done!

**Comment by sgnn7** [comment]
CR Pass

**Pull request merged**